ipn/ipnlocal/local.go

```diff
@@ -4451,9 +4457,30 @@ func (b *LocalBackend) stateMachine() {
 // around now and defer unlock in the caller to avoid missing unlocks and double
 // unlocks. TODO(bradfitz,maisem): make the locking in this package more
 // traditional (simple). See https://github.com/tailscale/tailscale/issues/11649
-func (b *LocalBackend) lockAndGetUnlock() (unlock func()) {
+func (b *LocalBackend) lockAndGetUnlock() (unlock unlockOnce) {
 	b.mu.Lock()
-	return sync.OnceFunc(b.mu.Unlock)
+	var unlocked atomic.Bool
+	return func() bool {
+		if unlocked.CompareAndSwap(false, true) {
+			b.mu.Unlock()
+			return true
+		}
+		return false
+	}
+}
+
+// unlockOnce is a func that unlocks only b.mu the first time it's called.
+// Therefore it can be safely deferred to catch error paths, without worrying
+// about double unlocks if a different point in the code later needs to explicitly
+// unlock it first as well. It reports whether it was unlocked.
+type unlockOnce func() bool
+
+// Unlock unlocks the LocalBackend.mu.
```

Comment:
This needs more docs. Or at least a different name, like UnlockEarly

Reply:
Like what property does it provide by panicing

Reply:
Documented, renamed.